indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/Acl.java

```diff
@@ -416,6 +427,37 @@ public static Principal getUserPrincipal(String userId, String identitySourceId)
         .setUserResourceName(String.format(USER_RESOURCE_NAME_FORMAT, identitySourceId, userId));
   }
 
+  private static Supplier<Set<String>> externalGroupNamesSupplier = () -> {
+    try {
+      ExternalGroups groups = ExternalGroups.fromConfiguration();
+      return groups.getExternalGroups().stream()
+          .map(group -> group.getName())
+          .collect(ImmutableSet.toImmutableSet());
+    } catch (IOException e) {
+      log.log(Level.INFO, "Not mapping external group names", e);
+      return ImmutableSet.of();
+    }
+  };
+
+  private static Supplier<Set<String>> externalGroupNames = memoize(externalGroupNamesSupplier);
+
+  private static Supplier<String> externalGroupsIdentitySourceIdSupplier = () -> {
+    return Configuration.getString("externalgroups.identitySourceId", "").get();
+  };
+
+  private static Supplier<String> externalGroupsIdentitySourceId =
+      memoize(externalGroupsIdentitySourceIdSupplier);
+
+  /** TestRule to reset the static cached external groups data for tests. */
+  public static class ResetExternalGroupsRule implements TestRule {
+    @Override
+    public Statement apply(Statement base, Description description) {
+      externalGroupsIdentitySourceId = memoize(externalGroupsIdentitySourceIdSupplier);
```

Comment:
One of these should be externalGroupNames? Also add a TODO about synchronization here.